### PR TITLE
Created a twistd plugin to run the tunnel helpers from.

### PR DESCRIPTION
**Pull request by lfdversluis**
Removed a test that ran the tunnel helper as the main.py file in the tunnel community no longer exists.

Adds the functionality of running it as a Twistd pluhin (feature).

### CODE REVIEW

**Comment by lfdversluis** [comment]
@devos50 can you review please?

File: twisted/twisted/plugins/tunnel_helper_plugin.py [hunk 1]
```diff
@@ -130,9 +162,7 @@ def start_tribler(self):
         config.set_enable_torrent_search(False)
         config.set_enable_channel_search(False)
         self.session = Session(config)
-        upgrader = self.session.prestart()
-        while not upgrader.is_done:
-            time.sleep(0.1)
+        self.session.prestart()
```
Comment (devos50) [inline twisted/twisted/plugins/tunnel_helper_plugin.py]: You should still check whether the upgrader failed or not. I did something like this:

```
upgrader = self.session.prestart()
if upgrader.failed:
    msg("The upgrader failed: .Tribler directory backed up, aborting")
    reactor.addSystemEventTrigger('after', 'shutdown', os._exit, 1)
    reactor.stop()
else:
    self.session.start()
    msg("Tribler started")
```

File: twisted/twisted/plugins/tunnel_helper_plugin.py [hunk 2]
```diff
@@ -337,69 +360,102 @@ def cb(ds):
...
+        self._stopping = False
+
+    def shutdown_process(self, shutdown_message, code=1):
+        msg(shutdown_message)
+        reactor.addSystemEventTrigger('after', 'shutdown', os._exit, code)
+        reactor.stop()
+
+    def start_tunnel(self, options):
+        """
+        Main method to startup a tunnel helper and add a signal handler..
```
Comment (devos50) [inline twisted/twisted/plugins/tunnel_helper_plugin.py]: This sentence should only end with one dot I think?

File: twisted/twisted/plugins/tunnel_helper_plugin.py [hunk 3]
```diff
@@ -337,69 +360,102 @@ def cb(ds):
...
+        profile = options["yappi"]
+
+        if profile:
+            yappi.set_clock_type(profile)
+            yappi.start(builtins=True)
+            logger.error("Profiling using %s time" % yappi.get_clock_type()['type'])
+
+        if crawl_keypair_filename and not os.path.exists(crawl_keypair_filename):
+            logger.error("Could not find keypair filename", crawl_keypair_filename)
+            sys.exit(1)
```
Comment (devos50) [inline twisted/twisted/plugins/tunnel_helper_plugin.py]: Maybe we should check this before jumping to the reactor thread?

**Comment by devos50** [comment]
Maybe you should check whether the passed ports (Dispersy/manhole) are valid?

File: twisted/twisted/plugins/tunnel_helper_plugin.py (hunk 1 above)
Comment (lfdversluis) [inline twisted/twisted/plugins/tunnel_helper_plugin.py]: This would not be enough, the upgrader opens the database which needs to be closed I think.

File: twisted/twisted/plugins/tunnel_helper_plugin.py (hunk 3 above)
Comment (lfdversluis) [inline twisted/twisted/plugins/tunnel_helper_plugin.py]: Hmmm may not be a bad idea!

**Comment by lfdversluis** [comment]
Ahh looks like GitHub is stable again.

**Comment by lfdversluis** [comment]
@whirm Processed the feedback, can you review this please?

**Comment by lfdversluis** [comment]
retest this please

File: twisted/twisted/plugins/tunnel_helper_plugin.py [hunk 4]
```diff
@@ -337,69 +365,131 @@ def cb(ds):
...
+            logger.error("Invalid socks5 port", socks5_port)
+            return False
+
+        introduce_port = options["introduce"]
+        if introduce_port and introduce_port <= 0:
+            logger.error("Invalid introduction port", introduce_port)
+            return False
+
+        dispersy_port = options["dispersy"]
+        if dispersy_port and (dispersy_port  <= 0 and dispersy_port != -1):
```
Comment (whirm) [inline twisted/twisted/plugins/tunnel_helper_plugin.py]: just do `< -1`

File: twisted/twisted/plugins/tunnel_helper_plugin.py [hunk 5]
```diff
@@ -337,69 +365,131 @@ def cb(ds):
...
+        Initialize the variables of this service and the logger.
+        """
+        self._stopping = False
+
+    def shutdown_process(self, shutdown_message, code=1):
+        msg(shutdown_message)
+        reactor.addSystemEventTrigger('after', 'shutdown', os._exit, code)
+        reactor.stop()
+
+    def check_options(self, options):
```
Comment (whirm) [inline twisted/twisted/plugins/tunnel_helper_plugin.py]: Integrate the option validation into the Option class: https://twistedmatrix.com/documents/12.0.0/core/howto/options.html

File: twisted/twisted/plugins/tunnel_helper_plugin.py [hunk 6]
```diff
@@ -256,14 +282,14 @@ def lineReceived(self, line):
             for circuit_id, circuit in anon_tunnel.community.circuits.items():
                 info_hash = circuit.info_hash.encode('hex')[:10] if circuit.info_hash else '?'
                 print "%d\t%s:%d\t%d\t%d\t\t%.2f\t\t%.2f\t\t%s\t%s" % (circuit_id,
```
Comment (whirm) [inline twisted/twisted/plugins/tunnel_helper_plugin.py]: why is this printing instead of logging?

**Comment by whirm** [comment]
Looks good!

File: twisted/twisted/plugins/tunnel_helper_plugin.py (hunk 6 above)
Comment (lfdversluis) [inline twisted/twisted/plugins/tunnel_helper_plugin.py]: Legacy code, will change.

File: twisted/twisted/plugins/tunnel_helper_plugin.py (hunk 4 above)
Comment (lfdversluis) [inline twisted/twisted/plugins/tunnel_helper_plugin.py]: is 0 allowed as port?

File: twisted/twisted/plugins/tunnel_helper_plugin.py (hunk 6 above)
Comment (whirm) [inline twisted/twisted/plugins/tunnel_helper_plugin.py]: Can't you make it a log call?

Comment (lfdversluis) [inline twisted/twisted/plugins/tunnel_helper_plugin.py]: Yes that's what I meant with will change. I will make use the logger :+1:

**Comment by lfdversluis** [comment]
retest this please

**Comment by lfdversluis** [comment]
Sorry @whirm, I did an --amend because I thought I had a fixup commit but that one was gone because I had to rebase because of #2151. So it is one commit now again.

**Comment by whirm** [comment]
retest this please

**Comment by synctext** [comment]
retest this please

File: twisted/twisted/plugins/tunnel_helper_plugin.py [hunk 7]
```diff
@@ -53,7 +124,6 @@ def start_walking(self):
 
 
 class Tunnel(object):
-
     __single = None
```
Comment (whirm) [inline twisted/twisted/plugins/tunnel_helper_plugin.py]: It doesn't make sense to have the singleton stuff on a twistd plugin

Comment (whirm) [inline twisted/twisted/plugins/tunnel_helper_plugin.py]: I feel like this whole class should be elsewhere. It's something that can be used outside of the twistd plugin.

Comment (whirm) [inline twisted/twisted/plugins/tunnel_helper_plugin.py]: Let's do that in a separate PR so that it's easier to track the changes on git.

File: twisted/twisted/plugins/tunnel_helper_plugin.py [hunk 8]
```diff
@@ -333,77 +401,107 @@ def cb(ds):
...
+            settings.socks_listen_ports = [random.randint(1000, 65535) for _ in range(5)]
+
+        settings.become_exitnode = options["exit"]
+        if settings.become_exitnode:
+            logger.info("Exit-node enabled")
+        else:
+            logger.info("Exit-node disabled")
+
+        tunnel = Tunnel(settings, crawl_keypair_filename, dispersy_port)
+        StandardIO(LineHandler(tunnel, profile))
```
Comment (whirm) [inline twisted/twisted/plugins/tunnel_helper_plugin.py]: Have you checked that this LineHandler thing works when running the plugin in non-daemon mode?

Comment (lfdversluis) [inline twisted/twisted/plugins/tunnel_helper_plugin.py]: Yup, works. When I enter "p"  I get profiling disabled! with the logger. Speaking of profiling, would you mind running one of them with yappi enabled so I get a good profile output to analyse the tunnels with?

Comment (whirm) [inline twisted/twisted/plugins/tunnel_helper_plugin.py]: Sure, remind me of that when we do the deployment

**Comment by whirm** [comment]
Remove the singleton stuff and I guess we can merge it

**Comment by lfdversluis** [comment]
I renamed the `__single` to `__instance` because it is still used elsewhere (in the static method). Is this ok @whirm ?

**Comment by whirm** [comment]
The whole singleton thing should be gone, including the get_instance() classmethod.

**Comment by lfdversluis** [comment]
But the `TunnelCommunityCrawler` makes use of the instance for the stats handles. (see line 118)

**Comment by lfdversluis** [comment]
@whirm forgot to mention ^

**Comment by whirm** [comment]
Then it needs to be refactored to not do so.

Ugh, I just found a bunch of stuff that needs to be fixed on the Tunnel constructor too...

Just kill the renaming of the singleton attribute as it doesn't solve anything and I'll create an issue with the stuff that needs to be refactored.

**Comment by lfdversluis** [comment]
retest this please

**Comment by whirm** [comment]
retest this please

**Comment by whirm** [comment]
(I'm testing some fixes I made in gumby)

**Comment by whirm** [comment]
Timeout cloning the repo...
